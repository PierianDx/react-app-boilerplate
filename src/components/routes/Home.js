import React, { Component } from 'react';

class Home extends Component {
  state = {
    response: '',
    post: '',
    responseToPost: ''
  };

  componentDidMount() {
    this.callApi()
      .then(res => this.setState({ response: res.express }))
      .catch(err => console.log(err));
  }

  callApi = async () => {
    const response = await fetch('/api/hello');
    const body = await response.json();

    if (response.status !== 200) throw Error(body.message);
    return body;
  };

  handleSubmit = async ev => {
    ev.preventDefault();
    const response = await fetch('/api/world', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ post: this.state.post })
    });

    const body = await response.text();
    this.setState({ responseToPost: body });
  };

  render() {
    return (
      <section className="section">
        <div className="container">
          <div className="notification">
            <p className="content">
              <strong>Response from Server: </strong>
              {this.state.response}
            </p>
          </div>
          <form onSubmit={this.handleSubmit}>
            <h4 className="title">Post Message to Server</h4>
            <div className="field has-addons">
              <div className="control has-icons-left is-expanded">
                <input
                  type="text"
                  className="input"
                  placeholder="type a message here"
                  value={this.state.post}
                  onChange={e => this.setState({ post: e.target.value })}
                />
                <span className="icon is-small is-left">
                  <i className="fas fa-envelope" />
                </span>
              </div>
              <div className="control">
                <button className="button is-primary" type="submit">
                  Submit
                </button>
              </div>
            </div>
          </form>
          {this.state.responseToPost && (
            <div className="notification">
              <p className="content">
                <strong>Received</strong>: {this.state.responseToPost}
              </p>
            </div>
          )}
        </div>
      </section>
    );
  }
}

export default Home;
