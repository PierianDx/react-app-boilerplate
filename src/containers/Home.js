import React, { Component } from 'react';
import Notification from '../components/Notification';

class Home extends Component {
  state = {
    response: '',
    post: '',
    active: false,
    responseToPost: ''
  };

  componentDidMount() {
    this.callApi()
      .then(res => this.setState({ response: res.express }))
      .catch(err => console.log(err));
  }

  callApi = async () => {
    const response = await fetch('/api/hello');
    const body = await response.json();

    if (response.status !== 200) throw Error(body.message);
    return body;
  };

  handleClose = ev => {
    this.setState({
      active: false,
      responseToPost: ''
    });
  };

  handleSubmit = async ev => {
    ev.preventDefault();
    const response = await fetch('/api/world', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ post: this.state.post })
    });

    const body = await response.text();
    this.setState({
      active: true,
      responseToPost: body
    });
  };

  render() {
    const { active, responseToPost, post, response } = this.state;
    return (
      <section className="section App">
        <p className="title">Home Page</p>
        <div className="container margin-md">
          <div className="notification">{response}</div>
        </div>
        <form onSubmit={this.handleSubmit} className="margin-md">
          <div className="field">
            <label className="label">Post to Server</label>
          </div>
          <div className="field is-grouped">
            <p className="control has-icons-left is-expand">
              <input
                type="text"
                className="input"
                value={post}
                onChange={e => this.setState({ post: e.target.value })}
              />
              <span className="icon is-small is-left">
                <i className="fas fa-globe" />
              </span>
            </p>
            <p className="control">
              <button className="button is-primary" type="submit">
                Send Message
              </button>
            </p>
          </div>
        </form>
        <div className="container margin-md">
          {responseToPost && (
            <Notification active={active} onClose={this.handleClose}>
              <p>
                Received <strong>{responseToPost}</strong> back from server.
              </p>
            </Notification>
          )}
        </div>
      </section>
    );
  }
}

export default Home;
