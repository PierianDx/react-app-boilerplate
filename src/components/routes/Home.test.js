import React from 'react';
import { shallow, mount } from 'enzyme';
import waitUntil from 'async-wait-until';
import nock from 'nock';

import Home from './Home';

describe('Home', () => {
  beforeEach(() => {
    // Mock network responses from API calls
    nock(/.*/)
      .persist()
      // .log(console.log)
      .get('/api/hello')
      .reply(
        200,
        {
          express: 'Hello from Test'
        },
        {
          'Access-Control-Allow-Origin': '*',
          'Content-type': 'application/json'
        }
      )
      .post('/api/world')
      .reply(200, (uri, body) => body.post, {
        'Access-Control-Allow-Origin': '*',
        'Content-type': 'application/json'
      });
  });

  afterEach(() => {
    // Reset network mocks after each test
    nock.cleanAll();
  });

  it('should render correctly', () => {
    const comp = shallow(<Home />);
    expect(comp).toMatchSnapshot();
  });

  it('should update based on sent messages', async done => {
    let comp = shallow(<Home />);

    // Handle componentDidMount async call
    await waitUntil(() => comp.state('response') !== '');
    expect(comp.state('response')).toEqual('Hello from Test');

    // Handling user input field
    comp.find('input').simulate('change', {
      target: { value: 'testing' }
    });
    expect(comp.state('post')).toEqual('testing');

    // Handle form submit (async call to api)
    comp.find('form').simulate('submit', { preventDefault: () => {} });
    await waitUntil(() => comp.state('responseToPost') !== '');
    expect(comp.state('responseToPost')).toEqual('testing');

    done();
  });
});
