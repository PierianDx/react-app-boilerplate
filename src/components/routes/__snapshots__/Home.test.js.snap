// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly 1`] = `
<section
  className="section"
>
  <div
    className="container"
  >
    <div
      className="notification"
    >
      <p
        className="content"
      >
        <strong>
          Response from Server: 
        </strong>
      </p>
    </div>
    <form
      onSubmit={[Function]}
    >
      <h4
        className="title"
      >
        Post Message to Server
      </h4>
      <div
        className="field has-addons"
      >
        <div
          className="control has-icons-left is-expanded"
        >
          <input
            className="input"
            onChange={[Function]}
            placeholder="type a message here"
            type="text"
            value=""
          />
          <span
            className="icon is-small is-left"
          >
            <i
              className="fas fa-envelope"
            />
          </span>
        </div>
        <div
          className="control"
        >
          <button
            className="button is-primary"
            type="submit"
          >
            Submit
          </button>
        </div>
      </div>
    </form>
  </div>
</section>
`;
